File: 2018-19/contest4/prog/junior/prefix.py
import os
import sys

for N in range(5):
	tokens = [x for x in input().split()]
	st = [] # A list of two-item lists where the first element 'True' for operators and 'False' for operands
	for ch in tokens:
		if ch == '+' or ch == '-' or ch == '*' or ch == '@':
			st.append([True,ch])
			continue
		st.append([False,int(ch)])
		found = False
		n = len(st)
		while True:
			found = False
			n = len(st)
			if n >= 3 and st[n-3][0] and st[n-3][1] != '@' and not st[n-2][0] and not st[n-1][0]:
				# found: +/-/* operand operand
				res = 0
				if st[n-3][1] == '+':
					res = st[n-2][1] + st[n-1][1]
				if st[n-3][1] == '-':
					res = st[n-2][1] - st[n-1][1]
				if st[n-3][1] == '*':
					res = st[n-2][1] * st[n-1][1]
				st.pop()
				st.pop()
				st.pop()
				st.append([False,res])
				found = True
				n = len(st)
			if n >= 4 and st[n-4][0] and st[n-4][1] == '@' and not st[n-3][0] and not st[n-2][0] and not st[n-1][0]:
				# found: @ operand operand operand
				res = st[n-2][1] if st[n-3][1] > 0 else st[n-1][1]
				st.pop()
				st.pop()
				st.pop()
				st.pop()
				st.append([False,res])
				found = True
				n = len(st)
			if not found or n < 3: # continue evaluation for possible expressions
				break
	print(st[0][1])

File: 2018-19/contest4/prog/junior/prefix.java
import java.io.*;
import java.util.*;

public class prefix {
	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		for(int N = 0; N < 5; ++N) {
			String line = input.nextLine();
			ArrayList<Boolean> st0 = new ArrayList<Boolean>(); // true when the token is an operator
			ArrayList<Integer> st1 = new ArrayList<Integer>();
			for(int k = 0; k < line.length(); ++k) {
				char ch = line.charAt(k);
				if(ch == ' ')
					continue;
				if(ch == '+' || ch == '-' || ch == '*' || ch == '@') {
					st0.add(true);
					st1.add((int)ch);
					continue;
				}
				st0.add(false);
				st1.add(ch-'0');
				boolean found = false;
				int n = st1.size();
				do {
					found = false;
					n = st1.size();
					if(n >= 3 && st0.get(n-3) && st1.get(n-3) != (int)'@' && !st0.get(n-2) && !st0.get(n-1)) {
						// found: +/-/* operand operand
						int res = 0;
						char op = (char)st1.get(n-3).intValue();
						switch(op){
						case '+': res = st1.get(n-2) + st1.get(n-1); break;
						case '-': res = st1.get(n-2) - st1.get(n-1); break;
						case '*': res = st1.get(n-2) * st1.get(n-1); break;
						}
						for(int i = 0; i < 3; ++i) {
							st0.remove(st0.size()-1);
							st1.remove(st1.size()-1);
						}
						st0.add(false);
						st1.add(res);
						found = true;
						n = st1.size();
					}
					if(n >= 4 && st0.get(n-4) && st1.get(n-4) == (int)'@' && !st0.get(n-3) && !st0.get(n-2) && !st0.get(n-1)) {
						// found: @ operand operand operand
						int res = st1.get(n-3) > 0 ? st1.get(n-2) : st1.get(n-1);
						for(int i = 0; i < 4; ++i) {
							st0.remove(st0.size()-1);
							st1.remove(st1.size()-1);
						}
						st0.add(false);
						st1.add(res);
						found = true;
						n = st1.size();
					}
				} while(found && n >= 3); // continue evaluation for possible expressions
			}
			System.out.println(st1.get(st1.size()-1));
		}
	}
}
